Document terraform code structure (#72)
shalb/cluster.dev: docs/terraform-code-structure.md

> # Terraform modules structure

Reviewer: ```suggestion
# Terraform modules' structure
```

## Filenames <!-- omit in toc -->

> * [`0-init.tf`](#0-inittf)

Reviewer 2: Why you have this numbers in naming?  0,1,8,9?

Author: For sort logically. 

Imagine that you have many files in your module (simple case [exist now in terraform/aws/addons](https://github.com/shalb/cluster.dev/tree/df78211dff8de8f3ab31c7bea61f03b5827734a4/terraform/aws/addons))

```
View now       | View in IDE     | View in Github  |
-------------- | --------------- | --------------- |
argocd.yaml    | 0-init.tf       | README.md       |
init.tf        | 1-main.tf       | 0-init.tf       |
file.sh        | 2-too-logic.tf  | 1-main.tf       |
main.tf        | 3-more-logic.tf | 2-too-logic.tf  |
more-logic.tf  | 8-outputs.tf    | 3-more-logic.tf |
README.md      | 9-vars.tf       | 8-outputs.tf    |
outputs.tf     | argocd.yaml     | 9-vars.tf       |
too-logic.tf   | file.sh         | argocd.yaml     |
something.yaml | README.md       | file.sh         |
vars.tf        | something.yaml  | something.yaml  |
```

And we can compose files with logic (`main.tf`, `too-logic.tf` `more-logic.tf`) as it will be really executed/depends. Or as we'd like

For ex:
`1-main.tf`, `2-too-logic.tf`, `3-more-logic.tf`
`1-more-logic.tf`, `2-too-logic.tf`, `3-main.tf` 
`1-too-logic.tf`, `2-more-logic.tf`, `3-main.tf`

Reviewer 2: outputs.tf and variables.tf - is the standart ways across all modules in internet :).
I'd like to keep thisn naming as is.
So lets approve this without numbers in front.
Also for templates I thinks we should use dedicated /templates folder

* [`1-main.tf`](#1-maintf)
* [`8-outputs.tf`](#8-outputstf)
* [`9-vars.tf`](#9-varstf)
* [`README.md`](#readmemd)

> `.yaml`, `.sh` and other related to module files put directly to module folder.

Reviewer: ```suggestion
Put the `.yaml`, `.sh` and other module-related files directly to the module folder.
```

> Also, look our [style-guide](style-guide.md) for check other best practices.

Reviewer: ```suggestion
Also, look our [style-guide](style-guide.md) to check other best practices.
```

## `0-init.tf`

Sort `terraform_remote_state` alphabetically.

```terraform
#
# Init
#

terraform {

>   required_version = ">= X.Y.Z" # minimum required terraform version. Older - better.

Reviewer: ```suggestion
  required_version = ">= X.Y.Z" # minimum required Terraform version. Older versions are preferable. 
```

  required_providers {

>     _ = "~> X.Y" # used provider version, pin only MAJOR.MINOR. Newest - better.

Reviewer: ```suggestion
    _ = "~> X.Y" # version of the used provider, pin only MAJOR.MINOR. The newest version is preferable.
```

  }

  backend "_" {}
}

provider "_" {}

#
# Remote states for import
#

data "terraform_remote_state" "" {}

```

## `1-main.tf`

Sort in logic order. Sort alphabetically when possible.

```terraform

#
# Get/transform exit data
#

locals {}

data "_" "_" {}

resource "_" "_" {}

module "_" {}

#
# Create resources
#

resource "_" "_" {}

module "_" {}

```

## `8-outputs.tf`

Sort alphabetically.

```terraform
output "_" {
  value       = _
  description = ""
}

```

## `9-vars.tf`

Sort alphabetically.

```terraform
variable "_" {
  type = _

  description = "_"
}

```

## `README.md`

```md
# Module name

> Describe What this module do.

Reviewer: ```suggestion
Describe what the module does.
```
